cassiopeia/solver/dissimilarity_functions.py

```diff
@@ -239,6 +242,69 @@ def weighted_hamming_similarity(
     return d / num_present
 
 
+def exponential_negative_hamming_distance(
+    s1: List[int],
+    s2: List[int],
+    missing_state_indicator=-1,
+    weights: Optional[Dict[int, Dict[int, float]]] = None,
+) -> float:
+
+    """
+    Calculates the default affinity for SpectralNeighborJoiningSolver. This
+    simply returns exp(-d(i,j)) where d is the weighted_hamming_distance
+    function as above where no weights are passed in.  In other words, we
+    increment d by +2 if the states are different, +1 if one state is uncut and
+    the other is an indel, and +0 if the two states are identical. Then, we take
+    this total d and return exp(-d). Note that since d is a metric,
+    'exponential_negative_hamming_distance' is a multiplicative similarity
+    score, i.e. s(i, j) = s(i, k) * s(k, j) for k on the path between i and j.
+
+    Args:
+        s1: Character states of the first sample
+        s2: Character states of the second sample
+        missing_state_indicator: The character representing missing values
+        weights: A dictionary storing the state weights for each character,
+            derived from the state priors. This should be a nested
+            dictionary where each key corresponds to character that then indexes
+            another dictionary storing the weight of each observed state.
+            (Character -> State -> Weight)
+
+    Returns:
+        A similarity score.
+    """
+
+    d = 0
+    num_present = 0
+    for i in range(len(s1)):
+
+        if s1[i] == missing_state_indicator or s2[i] == missing_state_indicator:
+            continue
+
+        num_present += 1
+
+        if s1[i] != s2[i]:
+            if s1[i] == 0 or s2[i] == 0:
+                if weights:
+                    if s1[i] != 0:
+                        d += weights[i][s1[i]]
+                    else:
+                        d += weights[i][s2[i]]
+                else:
+                    d += 1
+            else:
+                if weights:
+                    d += weights[i][s1[i]] + weights[i][s2[i]]
+                else:
+                    d += 2
+
+    if num_present == 0:
+        return 0
+
+    weighted_hamm_dist = d / num_present
```

Comment:
This function looks identical to `weighted_hamming_distance`. Why not just call this function, and then pass it to `np.exp(-val)`?

Reply:
so I deleted this and just called the function and it stopped working. Ivan said he did this because apparently calling external functions messes with numba? im not sure... temporarily leaving it hard coded

Reply:
I think Numba is called on the disimilarity function, and if I call weighted_hamming_distance inside our function, then we have to make weighted_hamming_distance compile with numba, which should be fine except it throws a lot of compilation errors. Somehow if I paste the weighted_hamming_distance  code into our function, and not use the @jit decorator inside the dissimilarity_functions.py, it works.

Reply:
Thanks for the info! I'll take a look at getting this to work.

Reply:
You're right - I can't get this to work because the function inside a function wrapped with `jit` must be `jit` compiled as well. Currently, our abstraction set up doesn't support this, so we'll have to live with duplicated code.